Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide/Internal.hs

```diff
@@ -110,67 +110,71 @@ endRideTransaction driverId bookingId ride mbFareParams mbRiderDetailsId merchan
+    Hedis.zAddExp (makeDailyDriverLeaderBoardKey merchantId rideDate) ride.driverId.getId (fromIntegral currZscore) lbExpiry.getSeconds
+    let limit = dailyLeaderBoardConfig.leaderBoardLengthLimit
+    driversListWithScores' <- Hedis.zrevrangeWithscores (makeDailyDriverLeaderBoardKey merchantId rideDate) 0 (limit -1)
+    Hedis.setExp (makeCachedDailyDriverLeaderBoardKey merchantId rideDate) driversListWithScores' (dailyLeaderBoardConfig.leaderBoardExpiry.getSeconds * dailyLeaderBoardConfig.numberOfSets)
+  pure ()
+
+updateDriverWeeklyZscore :: (Esq.EsqDBFlow m r, Esq.EsqDBReplicaFlow m r, Metrics.CoreMetrics m, CacheFlow m r, Hedis.HedisFlow m r, MonadFlow m) => Ride.Ride -> Day -> Day -> Day -> Maybe Double -> Maybe Meters -> Id Merchant -> m ()
+updateDriverWeeklyZscore ride rideDate weekStartDate weekEndDate driverZscore rideChargeableDistance merchantId = do
+  weeklyLeaderBoardConfig <- QLeaderConfig.findLeaderBoardConfigbyType LConfig.WEEKLY merchantId >>= fromMaybeM (InternalError "Leaderboard configs not present")
+  when weeklyLeaderBoardConfig.isEnabled $ do
+    (LocalTime _ localTime) <- utcToLocalTime timeZoneIST <$> getCurrentTime
+    let (_, currDayIndex) = sundayStartWeek rideDate
+    let lbExpiry = weeklyLeaderBoardConfig.leaderBoardExpiry - Seconds ((currDayIndex + 1) * 86400) + (Seconds 86400 - secondsFromTimeOfDay localTime) -- Calculated as (total_Seconds_in_week - Week_Day_Index * 86400 + Seconds_Remaining_In_This_Week
+    let currZscore =
+          case driverZscore of
+            Nothing -> weeklyLeaderBoardConfig.zScoreBase + getMeters (fromMaybe 0 rideChargeableDistance)
+            Just zscore -> do
+              let (prevTotalRides, prevTotalDistance) = getRidesAndDistancefromZscore zscore weeklyLeaderBoardConfig.zScoreBase
+              let currTotalRides = prevTotalRides + 1
+              let currTotalDist = prevTotalDistance + fromMaybe 0 rideChargeableDistance
+              currTotalRides * weeklyLeaderBoardConfig.zScoreBase + getMeters currTotalDist
+    Hedis.zAddExp (makeWeeklyDriverLeaderBoardKey merchantId weekStartDate weekEndDate) ride.driverId.getId (fromIntegral currZscore) (lbExpiry.getSeconds)
+    let limit = weeklyLeaderBoardConfig.leaderBoardLengthLimit
+    driversListWithScores' <- Hedis.zrevrangeWithscores (makeWeeklyDriverLeaderBoardKey merchantId weekStartDate weekEndDate) 0 (limit -1)
+    Hedis.setExp (makeCachedWeeklyDriverLeaderBoardKey merchantId weekStartDate weekEndDate) driversListWithScores' (weeklyLeaderBoardConfig.leaderBoardExpiry.getSeconds * weeklyLeaderBoardConfig.numberOfSets)
+  pure ()
+
+makeCachedDailyDriverLeaderBoardKey :: Id Merchant -> Day -> Text
+makeCachedDailyDriverLeaderBoardKey merchantId rideDate = "DDLBCK:" <> show merchantId <> ":" <> show rideDate
+
+makeDailyDriverLeaderBoardKey :: Id Merchant -> Day -> Text
+makeDailyDriverLeaderBoardKey merchantId rideDate =
+  "DDLBK:" <> show merchantId <> ":" <> show rideDate
```

Comment:
you can do `merchantId.getId` ? show will give you -> `Id { getId = "<uuid>" }` as a string.